File: t2/open-source/bits-and-bobs/square-waves-decomposition/square-waves-decompose-0.1.0.pl
#!/usr/bin/perl
#
# square-waves-decompose.pl : compose a sine wave out of square waves.
#
# The results of this program can be plotted using the gnuplot command:
#
#     plot "gnuplot.dat" , "sinplot.dat"
#
# Written by Shlomi Fish, 2008.
#
# Licensed under the MIT X11 license:
# http://www.opensource.org/licenses/mit-license.php )
#
# Requires:
# * perl-5.8.x (may work with older versions)
# * The List::Util module.
# * The PDL module: http://pdl.perl.org/
#
# Version: 0.1.0

use strict;
use warnings;

use POSIX ();
use PDL;
use List::Util ();

use constant PI => (4 * atan2(1, 1));

sub square_wave
{
    my ($len, $height, $x_s) = @_;

    return [map {(POSIX::floor($_*$len)%2 == 0) ? $height : -$height } @$x_s];
}

sub primes
{
    my $n = shift;

    my @ret = (2);
    my $next = 3;
    MAIN:
    while (@ret != $n)
    {
        foreach (@ret)
        {
            if ($next % $_ == 0)
            {
                next MAIN;
            }
        }
    }
}

my $num = 100;
my @points = (map { $_ / $num } (-$num .. $num));

my $good = pdl([map { sin($_ * PI()) } @points]);

# my @lengths = ();

my $len = 1;

my @waves;

my $energy = ($good*$good)->average();
while ($energy > 0.0002)
{
    my $base = pdl(square_wave($len, 1, \@points));

    # The co-efficient to multiply
    # If we want min ||y-> - m*x->|| then we need to have
    # m = Sum[xy]/Sum[x^2]
    my $m = ($good * $base)->sum() / ($base*$base)->sum();
    push @waves, ([$len, $m]);
    $good -= $base*$m;
}
continue
{
    $energy = ($good*$good)->average();
    print "$energy\n";
    $len++;
}

open my $plot, ">", "gnuplot.dat";
my $x_num = 10_000;
foreach my $x_int (-$x_num .. $x_num)
{
    my $x = $x_int / $x_num;
    printf {$plot} ("%.10f %.10f\n", $x,
        List::Util::sum(
            map { @{square_wave(@$_, [$x])} } @waves
        ));
}
close($plot);

if (! -e "sinplot.dat")
{
open my $sin_plot, ">", "sinplot.dat";
foreach my $x_int (-10_000 .. 10_000)
{
    my $x = $x_int / 10_000;
    printf {$sin_plot} ("%.10f %.10f\n" , $x, sin($x*PI()));
}
close($sin_plot);
}

File: t2/open-source/bits-and-bobs/square-waves-decomposition/square-waves-decompose-0.1.0.c
/*
 * square-waves.c : compose a sine wave out of square waves.
 *
 * the results of this program can be plotted using the gnuplot command:
 *
 *      plot "gnuplot.dat" , "sinplot.dat"
 *
 * Written by Shlomi Fish, 2007.
 *
 * Licensed under the MIT X11 license.
 *
 * Compile with:
 * gcc -o square-waves -Wall square-waves.c -lm
 *
 * Version: 0.1.0
 *
 * */
#include <stdio.h>
#include <math.h>
#include <stdlib.h>
#include <unistd.h>

typedef double value_t;

struct wave_struct
{
    /* The length of the wave. */
    int len;
    /* Its amplitude */
    value_t m;
};

typedef struct wave_struct wave_t;

#define CALC_SQUARE_WAVE(len,height,x) \
    ((((int)floor((x) * (len)))%2 == 0) ? (height) : (-(height)))

void square_wave(const value_t len, const value_t height,
        const int num_points, const value_t const * x_s, value_t * y_s
        )
{
    int i;

    for(i=0; i<num_points; i++)
    {
        y_s[i] = CALC_SQUARE_WAVE(len, height, x_s[i]);
    }
}

value_t calc_average(value_t * points, int num_points)
{
    value_t sum = 0;
    int i;

    for(i = 0 ; i < num_points ; i++)
    {
        sum += points[i]*points[i];
    }

    return sum/num_points;
}

int main(int argc, char * argv[])
{
    int num = 100;
    int x_num = 10000;
    int num_points;
    int i;
    int len;
    int x_int;

    value_t * good, * points, * p_ptr, * good_ptr, * base;
    value_t energy;
    wave_t * waves;

    int num_waves = 0, max_num_waves = 0;

    FILE * plot_fh, * sin_fh;

    num_points = num*2+1;

    points = malloc(sizeof(points[0]) * num_points);
    good = malloc(sizeof(good[0]) * num_points);
    base = malloc(sizeof(base[0]) * num_points);

    waves = malloc(sizeof(waves[0]) * max_num_waves);

    /*
     * Initialize points to -num ... num
     * Initialise good to the sinousidial wave.
     * */
    for(i=-num, p_ptr = points, good_ptr = good;
            i<=num;
            i++, p_ptr++, good_ptr++
       )
    {
        *p_ptr = (value_t)i / num;
        *good_ptr = sin(*p_ptr * M_PI);
    }

    len = 1;
    energy = calc_average(good, num_points);
    while (energy > 0.0002)
    {
        value_t numer = 0, denom = 0, m;

        square_wave((value_t)len, 1, num_points, points, base);

        for(i=0;i<num_points;i++)
        {
            numer += good[i] * base[i];
            denom += base[i] * base[i];
        }

        m = numer / denom;

        if (num_waves == max_num_waves)
        {
            max_num_waves += 16;
            waves = realloc(waves, sizeof(waves[0])*max_num_waves);
        }
        waves[num_waves].len = len;
        waves[num_waves].m = m;

        num_waves++;

        for(i=0;i<num_points;i++)
        {
            good[i] -= base[i] * m;
        }

        energy = calc_average(good, num_points);
        printf("%f\n", energy);
        len++;
    }

    free(base);

    plot_fh = fopen("gnuplot.dat", "wt");

    for(x_int=-x_num;x_int<=x_num;x_int++)
    {
        value_t x = ((value_t)x_int) / x_num;
        int wave_idx;
        value_t result;

        result = 0.0;
        for(wave_idx = 0 ; wave_idx < num_waves ; wave_idx++)
        {
            result +=
                CALC_SQUARE_WAVE(waves[wave_idx].len, waves[wave_idx].m, x)
                ;
        }
        fprintf(plot_fh, "%.10f %.10f\n", x, result);
    }
    fclose(plot_fh);

    if (!!access("sinplot.dat", F_OK))
    {
        sin_fh = fopen("sinplot.dat", "wt");
        for(x_int=-x_num;x_int<=x_num;x_int++)
        {
            value_t x = ((value_t)x_int) / x_num;

            fprintf(sin_fh, "%.10f %.10f\n", x, sin(x*M_PI));
        }
        fclose(sin_fh);
    }

    free(points);
    free(good);
    free(waves);

    return 0;
}
